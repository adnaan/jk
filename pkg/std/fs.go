package std

import (
	"os"

	"github.com/justkidding-config/jk/pkg/__std"

	flatbuffers "github.com/google/flatbuffers/go"
)

func fileInfo(path string) []byte {
	info, err := os.Stat(path)
	switch {
	case err != nil:
		return fsError(err.Error())
	case !(info.IsDir() || info.Mode().IsRegular()):
		return fsError("not a regular file")
	}
	return fileInfoResponse(path, info.IsDir())
}

func fsError(msg string) []byte {
	b := flatbuffers.NewBuilder(1024)
	off := b.CreateString(msg)
	__std.ErrorStart(b)
	__std.ErrorAddMessage(b, off)
	off = __std.ErrorEnd(b)
	__std.FileSystemResponseStart(b)
	__std.FileSystemResponseAddRetvalType(b, __std.FileSystemRetvalError)
	__std.FileSystemResponseAddRetval(b, off)
	off = __std.FileSystemResponseEnd(b)
	b.Finish(off)
	return b.FinishedBytes()
}

func fileInfoResponse(path string, isdir bool) []byte {
	b := flatbuffers.NewBuilder(1024)
	off := buildFileInfo(b, path, isdir)
	__std.FileSystemResponseStart(b)
	__std.FileSystemResponseAddRetvalType(b, __std.FileSystemRetvalFileInfo)
	__std.FileSystemResponseAddRetval(b, off)
	off = __std.FileSystemResponseEnd(b)
	b.Finish(off)
	return b.FinishedBytes()
}

func buildFileInfo(b *flatbuffers.Builder, path string, isdir bool) flatbuffers.UOffsetT {
	off := b.CreateString(path)
	__std.FileInfoStart(b)
	__std.FileInfoAddPath(b, off)
	if isdir {
		__std.FileInfoAddIsdir(b, 1)
	} else {
		__std.FileInfoAddIsdir(b, 0)
	}
	return __std.FileInfoEnd(b)
}

func directoryListing(path string) []byte {
	dir, err := os.Open(path)
	if err != nil {
		return fsError(err.Error())
	}
	infos, err := dir.Readdir(0)
	if err != nil {
		return fsError(err.Error())
	}

	b := flatbuffers.NewBuilder(1024)
	offsets := make([]flatbuffers.UOffsetT, len(infos), len(infos))
	for i := range infos {
		offsets[i] = buildFileInfo(b, infos[i].Name(), infos[i].IsDir())
	}

	__std.DirectoryStartFilesVector(b, len(offsets))
	for i := len(offsets) - 1; i >= 0; i-- {
		b.PrependUOffsetT(offsets[i])
	}
	infoVec := b.EndVector(len(offsets))

	pathOff := b.CreateString(path)
	__std.DirectoryStart(b)
	__std.DirectoryAddPath(b, pathOff)
	__std.DirectoryAddFiles(b, infoVec)
	dirOff := __std.DirectoryEnd(b)

	__std.FileSystemResponseStart(b)
	__std.FileSystemResponseAddRetvalType(b, __std.FileSystemRetvalDirectory)
	__std.FileSystemResponseAddRetval(b, dirOff)
	off := __std.FileSystemResponseEnd(b)
	b.Finish(off)
	return b.FinishedBytes()
}
