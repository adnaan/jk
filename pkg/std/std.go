package std

import (
	"log"

	"github.com/dlespiau/jk/pkg/__std"

	flatbuffers "github.com/google/flatbuffers/go"
)

// Execute parses a message from v8 and execute the corresponding function.
func Execute(msg []byte) []byte {
	message := __std.GetRootAsMessage(msg, 0)

	union := flatbuffers.Table{}
	if !message.Args(&union) {
		log.Fatal("could not parse Message")
	}

	switch message.ArgsType() {
	case __std.ArgsWriteArgs:
		args := __std.WriteArgs{}
		args.Init(union.Bytes, union.Pos)
		write(args.Value(), string(args.Path()), args.Type())
		return nil
	default:
		log.Fatalf("unknown Message (%d)", message.ArgsType())
		return nil
	}
}
