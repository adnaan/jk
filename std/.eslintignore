flatbuffers.js
flatbuffers.d.ts
__std*.js
__std*.ts
