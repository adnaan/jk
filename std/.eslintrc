{
  "extends": "airbnb-base",
  "plugins": ["eslint-plugin-jest"],
  "globals": {
    "V8Worker2": true
  },
  "settings": {
     "import/resolver": {
        "node": {
          "paths": ["."]
        }
     }
  },
  "env": {
    "jest/globals": true
  },
  "rules": {
    "no-continue": 0,
    "prefer-const": ["error", {"destructuring": "all"}],
    "import/prefer-default-export": 0,
    "import/no-unresolved": [
      2,
      {
        "ignore": [
          "^@jkcfg/std$"
        ]
      }
    ],
    "indent": ["error", 2, {"SwitchCase": 0, "CallExpression": {"arguments": "first"}}],
    "no-restricted-syntax": ["error", "ForInStatement", "LabeledStatement", "WithStatement"],
    "no-use-before-define": ["error", { "functions": false }]
  }
}
