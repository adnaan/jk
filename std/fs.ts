/**
 * @module std/fs
 */

import { RPCSync } from './internal/rpc';
import { valueFromUTF8Bytes } from './internal/data';

export class FileInfo {
  name: string;
  path: string;
  isdir: boolean;

  constructor(n: string, p: string, d: boolean) {
    this.name = n;
    this.path = p;
    this.isdir = d;
  }
}

export class Directory {
  name: string;
  path: string;
  files: FileInfo[];

  constructor(n: string, p: string, files: FileInfo[]) {
    this.name = n;
    this.path = p;
    this.files = files;
  }
}

export interface InfoOptions {
  module?: string;
}

export function info(path: string, options: InfoOptions = {}): FileInfo {
  const { module = '' } = options;
  const response = RPCSync('std.fileinfo', path, module);
  const { name: n, path: p, isdir: d } = valueFromUTF8Bytes(response);
  return new FileInfo(n, p, d);
}

export interface DirOptions {
  module?: string;
}

export function dir(path: string, options: DirOptions = {}): Directory {
  const { module = '' } = options;
  const response = RPCSync('std.dir', path, module);
  const { name: n, path: p, files: fs } = valueFromUTF8Bytes(response);
  const infos = [];
  for (const f of fs) {
    const { name: infoname, path: infopath, isdir: infoisdir } = f;
    infos.push(new FileInfo(infoname, infopath, infoisdir));
  }
  return new Directory(n, p, infos);
}

export function join(base: string, name: string): string {
  return `${base}/${name}`;
}

const always = (): boolean => true;
const noop = (): void => {};

export interface WalkOpts {
  pre?: (f: FileInfo) => boolean;
  post?: () => void;
}

/* eslint-disable no-labels */

/** walk is a generator function that yields the files and directories
 * under a given path, in a preorder traversal. "Preorder" means that
 * the traversal is depth-first, and a directory is yielded
 * immediately before the traversal examines its contents.
 *
 * @param path the starting point for the traversal, which should be a
 * directory.
 *
 * @param opts pre- and post-hooks for the walk. The pre-hook is
 * called for each directory after it is yielded; if it returns a
 * falsey value, the directory is not traversed. The post-hook is
 * called after a directory's contents have all been traversed. The
 * starting point does not get treated as part of the traversal, i.e.,
 * it starts with the contents of the directory at `path`.
*/
export function* walk(path: string, opts: WalkOpts = {}): IterableIterator<FileInfo> {
  const { pre = always, post = noop } = opts;
  const top = dir(path);
  // the stack is going to keep lists of files to examine
  const stack: FileInfo[][] = [];
  let next = top.files;

  // eslint-disable-next-line no-restricted-syntax
  runNext: while (next !== undefined) {
    let i = 0;
    for (; i < next.length; i += 1) {
      const f = next[i];
      yield f;
      if (f.isdir && pre(f)) {
        const d = dir(f.path);
        // If we need to recurse into the subdirectory, push the work
        // yet to do here, then process the subdirectory's files.
        if (d.files.length > 0) {
          stack.push(next.slice(i + 1));
          next = d.files;
          continue runNext;
        }
        // If not, we can just continue through the current directory.
        post();
      }
    }
    // If we've exhausted the slice, we're popping a directory
    if (i === next.length && stack.length > 0) post();
    next = stack.pop();
  }
}

/* eslint-enable */
