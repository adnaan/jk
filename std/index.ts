export { log } from './log';
export {
  stdout,
  Format,
  Overwrite,
  write,
  print,
} from './write';
export { Encoding, read, stdin } from './read';
export { parse, stringify } from './parse';
