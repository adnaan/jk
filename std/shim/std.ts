function log(value: any): void { };

enum Format {
    JSON = 1,
    YAML,
    Raw,
}

interface WriteOptions {
  format?: Format,
  indent?: number,
  override?: boolean,
}

function write(value: any, path: string, options: WriteOptions): void { };

enum Encoding {
  Bytes,
  String,
  JSON,
}

interface ReadOptions {
  format: Format,
  encoding: Encoding,
}

function read(path: string, options: ReadOptions): Promise<ArrayBuffer | string | object> { return Promise.resolve({}) };

export default {
  log, Format, write, read,
};
