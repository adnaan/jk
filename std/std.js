import { log } from 'std_log';
import { Format, write } from 'std_write';
import { Encoding, read } from 'std_read';
import { info, dir } from 'std_fs';
import { param } from 'std_param';
import { mix, patch, merge } from 'std_merge';

// The default export is deprecated and will be removed in 3.0.0.

export default {
  log,
  Format,
  write,
  Encoding,
  read,
  info,
  dir,
  param,
  mix,
  patch,
  merge,
};

export { log } from 'std_log';
export { Format, write } from 'std_write';
export { Encoding, read } from 'std_read';
export { info, dir } from 'std_fs';
export { param } from 'std_param';
export { mix, patch, merge } from 'std_merge';
