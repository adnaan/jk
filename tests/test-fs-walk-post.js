import { walk } from '@jkcfg/std/fs';
import { print } from '@jkcfg/std';

// walk should call post for every time it finishes a directory. to
// check that the hook is called at the right time, include it in the
// transcript.

const post = () => print('post');
const pre = f => print(`pre ${f.name}`) || true;

for (const f of walk('./fs-walk-preorder-files', { pre, post })) {
  if (!f.name.startsWith('.')) print(f.name);
}
