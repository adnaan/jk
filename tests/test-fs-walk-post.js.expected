A
pre A
B
pre B
C
D
post
E
pre E
post
F
pre F
G
H
pre H
I
post
post
post
